File: javascript/868.js
/**
 * @param {number} n
 * @return {number}
 */
var binaryGap = function(n) {
    function isLastBitSet(n) {
        return (n & 1) == 1;
    };

    let count = -1;
    let best = 0;
    let num = n;

    while (num != 0) {
        if (isLastBitSet(num)) {
            if (count > best) {
                best = count;
            }

            count = 1;
        } else {
            if (count != -1) count += 1;
        }

        num = num >>> 1;
    };

    return best;
};

File: kotlin/868.kt
class Solution {
    fun binaryGap(n: Int): Int {
        if (n.countOneBits() == 0) return 0

        var bestDistance = 0
        var num = n
        var count = -1

        while (num != 0) {
            if (num.isLastBitSet()) {
                if (count > bestDistance) {
                    bestDistance = count
                }

                count = 1
            } else {
                if (count != -1) count += 1
            }

            num = num shr 1
        }

        return bestDistance
    }


    fun Int.isLastBitSet(): Boolean {
        return (this and 1) == 1
    }
}
